File: Euclid/Game/Room/Mapping/RoomFurniture.cs
using System;
using Euclid.Messages.Outgoing;
using Euclid.Storage.Database.Access;

namespace Euclid.Game
{
    public class RoomFurniture
    {
        #region Fields

        private Room room;

        #endregion

        #region Constructors

        public RoomFurniture(Room room)
        {
            this.room = room;
        }

        #endregion

        #region Public methods

        /// <summary>
        /// Add item to map handler
        /// </summary>
        internal void AddItem(Item item, Position position = null, string wallPosition = null, Player player = null)
        {
            item.Data.RoomId = room.Data.Id;

            if (item.Definition.Data.IsWallItem)
            {
                item.Data.WallPosition = wallPosition;
                room.Send(new ADDITEM(item));
            }
            else
            {
                RoomTile tile = position.GetTile(room);

                if (tile == null)
                    return;

                position.Z = tile.TileHeight;

                item.Data.X = position.X;
                item.Data.Y = position.Y;
                item.Data.Z = position.Z;
                item.Data.Rotation = position.Rotation;
                item.ApplyPosition();

                HandleItemAdjusted(item, false);

                room.Send(new ACTIVEOBJECT_ADD(item));
                room.Mapping.AddItem(item);
                item.UpdateEntities();    
            }

            room.ItemManager.AddItem(item);
            
            ItemDao.SaveItem(item.Data);
        }

        /// <summary>
        /// Move item handler
        /// </summary>
        internal void MoveItem(Item item, Position position = null, string wallPosition = null)
        {
            if (item.Definition.Data.IsWallItem)
            {
                item.Data.WallPosition = wallPosition;
            }
            else
            {
                bool isRotation = false;

                if (item.Position == new Position(position.X, position.Y) && item.Position.Rotation != position.Rotation)
                    isRotation = true;

                var oldTile = item.Position.GetTile(room);

                if (oldTile == null)
                    return;

                room.Mapping.RemoveItem(item);

                var newTile = position.GetTile(room);

                if (newTile == null)
                    return;

                position.Z = newTile.TileHeight;

                item.Data.X = position.X;
                item.Data.Y = position.Y;
                item.Data.Z = position.Z;
                item.Data.Rotation = position.Rotation;
                item.ApplyPosition();


                HandleItemAdjusted(item, isRotation);

                room.Send(new ACTIVEOBJECT_UPDATE(item));
                room.Mapping.AddItem(item);

                item.UpdateEntities(oldTile.Position);

            }
            
            ItemDao.SaveItem(item.Data);
        }

        private void HandleItemAdjusted(Item item, bool isRotation)
        {
            RoomTile tile = item.Position.GetTile(room);

            if (tile == null)
            {
                return;
            }

            if (!isRotation)
            {
                Item highestItem = tile.HighestItem;
                double tileHeight = tile.TileHeight;

                if (highestItem != null && highestItem.Data.Id == item.Data.Id)
                {
                    tileHeight -= highestItem.Height;

                    double defaultHeight = room.Model.TileHeights[item.Position.X, item.Position.Y];//this.room.getModel().getTileHeight(item.getPosition().getX(), item.getPosition().getY());

                    if (tileHeight < defaultHeight)
                        tileHeight = defaultHeight;
                }

                item.Position.Z = tileHeight;//.getPosition().setZ(tileHeight);
            }
        }


        /// <summary>
        /// Remove item handler
        /// </summary>
        public void RemoveItem(Item item, Player player)
        {
            room.Mapping.RemoveItem(item);

            if (item.Definition.Data.IsWallItem)
            {
                item.Data.WallPosition = string.Empty;
                room.Send(new REMOVEITEM(item));
            }
            else
            {
                item.UpdateEntities();

                item.Data.X = item.Position.X;
                item.Data.Y = item.Position.Y;
                item.Data.Z = item.Position.Z;
                item.Data.Rotation = item.Position.Rotation;
                item.ApplyPosition();

                room.Send(new ACTIVEOBJECT_REMOVE(item));
            }

            item.Data.RoomId = 0;

            room.ItemManager.RemoveItem(item);
            ItemDao.SaveItem(item.Data);
        }

        #endregion
    }
}

File: euclid/src/main/java/euclid/game/room/RoomFurniture.java
package euclid.game.room;

import euclid.game.Player;
import euclid.game.item.Item;
import euclid.game.pathfinder.Position;
import euclid.messages.outgoing.ACTIVEOBJECT_ADD;
import euclid.messages.outgoing.ACTIVEOBJECT_REMOVE;
import euclid.messages.outgoing.ACTIVEOBJECT_UPDATE;
import euclid.messages.outgoing.ADDITEM;
import euclid.messages.outgoing.REMOVEITEM;
import euclid.storage.database.access.ItemDao;

public class RoomFurniture {

    private final Room room;

    public RoomFurniture(Room room) {
        this.room = room;
    }

    public void addItem(Item item, Position position, String wallPosition, Player player) {
        item.getData().setRoomId(room.getData().getId());

        if (item.getDefinition().getData().isWallItem()) {
            item.getData().setWallPosition(wallPosition);
            room.send(new ADDITEM(item));
        } else {
            RoomTile tile = position.getTile(room);

            if (tile == null)
                return;

            position.setZ(tile.getTileHeight());

            item.getData().setX(position.getX());
            item.getData().setY(position.getY());
            item.getData().setZ(position.getZ());
            item.getData().setRotation(position.getRotation());
            item.applyPosition();

            handleItemAdjusted(item, false);

            room.send(new ACTIVEOBJECT_ADD(item));
            room.getMapping().addItem(item);
            item.updateEntities();
        }

        room.getItemManager().addItem(item);

        ItemDao.saveItem(item.getData());
    }

    public void addItem(Item item) {
        addItem(item, null, null, null);
    }

    public void moveItem(Item item, Position position, String wallPosition) {
        if (item.getDefinition().getData().isWallItem()) {
            item.getData().setWallPosition(wallPosition);
        } else {
            boolean isRotation = false;

            if (item.getPosition().equals(new Position(position.getX(), position.getY()))
                    && item.getPosition().getRotation() != position.getRotation())
                isRotation = true;

            RoomTile oldTile = item.getPosition().getTile(room);
            if (oldTile == null)
                return;

            room.getMapping().removeItem(item);

            RoomTile newTile = position.getTile(room);
            if (newTile == null)
                return;

            position.setZ(newTile.getTileHeight());

            item.getData().setX(position.getX());
            item.getData().setY(position.getY());
            item.getData().setZ(position.getZ());
            item.getData().setRotation(position.getRotation());
            item.applyPosition();

            handleItemAdjusted(item, isRotation);

            room.send(new ACTIVEOBJECT_UPDATE(item));
            room.getMapping().addItem(item);

            item.updateEntities(oldTile.getPosition());
        }

        ItemDao.saveItem(item.getData());
    }

    public void moveItem(Item item, Position position) {
        moveItem(item, position, null);
    }

    private void handleItemAdjusted(Item item, boolean isRotation) {
        RoomTile tile = item.getPosition().getTile(room);

        if (tile == null) {
            return;
        }

        if (!isRotation) {
            Item highestItem = tile.getHighestItem();
            double tileHeight = tile.getTileHeight();

            if (highestItem != null && highestItem.getData().getId() == item.getData().getId()) {
                tileHeight -= highestItem.getHeight();

                double defaultHeight = room.getModel().getTileHeights()[item.getPosition().getX()][item.getPosition().getY()];

                if (tileHeight < defaultHeight)
                    tileHeight = defaultHeight;
            }

            item.getPosition().setZ(tileHeight);
        }
    }

    public void removeItem(Item item, Player player) {
        room.getMapping().removeItem(item);

        if (item.getDefinition().getData().isWallItem()) {
            item.getData().setWallPosition("");
            room.send(new REMOVEITEM(item));
        } else {
            item.updateEntities();

            item.getData().setX(item.getPosition().getX());
            item.getData().setY(item.getPosition().getY());
            item.getData().setZ(item.getPosition().getZ());
            item.getData().setRotation(item.getPosition().getRotation());
            item.applyPosition();

            room.send(new ACTIVEOBJECT_REMOVE(item));
        }

        item.getData().setRoomId(0);

        room.getItemManager().removeItem(item);
        ItemDao.saveItem(item.getData());
    }
}
